src/backend/distributed/commands/multi_copy.c

```diff
@@ -2311,46 +2408,99 @@ CitusSendTupleToPlacements(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest
 	/* connections hash is kept in memory context */
 	MemoryContextSwitchTo(copyDest->memoryContext);
 
-	/* get existing connections to the shard placements, if any */
-	shardConnections = GetShardHashConnections(shardConnectionHash, shardId,
-											   &shardConnectionsFound);
-	if (!shardConnectionsFound)
+	shardState = GetShardState(shardId, copyDest->shardStateHash,
+							   copyDest->connectionStateHash, stopOnFailure,
+							   &cachedShardStateFound);
+	if (!cachedShardStateFound)
 	{
-		/*
-		 * Keep track of multi shard accesses before opening connection
-		 * the second shard.
-		 */
-		if (!copyDest->multiShardCopy && hash_get_num_entries(shardConnectionHash) == 2)
-		{
-			Oid relationId = copyDest->distributedRelationId;
+		firstTupleInShard = true;
+	}
+
+	if (firstTupleInShard && !copyDest->multiShardCopy &&
+		hash_get_num_entries(copyDest->shardStateHash) == 2)
+	{
+		Oid relationId = copyDest->distributedRelationId;
 
-			/* mark as multi shard to skip doing the same thing over and over */
-			copyDest->multiShardCopy = true;
+		/* mark as multi shard to skip doing the same thing over and over */
+		copyDest->multiShardCopy = true;
 
-			/* error out of conflicting COPY */
-			CheckConflictingParallelCopyAccesses(relationId);
+		/* error out of conflicting COPY */
+		CheckConflictingParallelCopyAccesses(relationId);
 
+		if (MultiShardConnectionType != SEQUENTIAL_CONNECTION)
+		{
 			/* when we see multiple shard connections, we mark COPY as parallel modify */
 			RecordParallelModifyAccess(relationId);
 		}
+	}
 
-		/* open connections and initiate COPY on shard placements */
-		OpenCopyConnections(copyStatement, shardConnections, stopOnFailure,
-							copyOutState->binary);
+	foreach(placementStateCell, shardState->placementStateList)
+	{
+		CopyPlacementState *placementState = lfirst(placementStateCell);
+		CopyConnectionState *connectionState = placementState->connectionState;
+		CopyPlacementState *currentActivePlacementState = connectionState->activePlacementState;
+		bool switchOver = false;
+		bool sendTupleOverConnection = false;
 
-		/* send copy binary headers to shard placements */
-		if (copyOutState->binary)
+		if (currentActivePlacementState == NULL)
 		{
-			SendCopyBinaryHeaders(copyOutState, shardId,
-								  shardConnections->connectionList);
+			switchOver = true;
+		}
+		else if (placementState != currentActivePlacementState &&
+				 placementState->data->len > COPY_SWITCH_OVER_THRESHOLD)
+		{
+			switchOver = true;
```

Comment:
To clarify what we are switching to, maybe we can call this `switchToTargetShard` and rename `shardState` to something like `targetShardState`?

Reply:
or `switchToCurrentShard` and `currentShardState` and rename `currentActivePlacementState` to `activePlacementState`.

Reply:
> or `switchToCurrentShard` and `currentShardState` and rename `currentActivePlacementState` to `activePlacementState`.

done.